File: 670-maximum-swap/maximum-swap.cpp
class Solution {
public:
    int maximumSwap(int num) {
        string s = to_string(num);
        sort(s.begin(), s.end(), greater<char>());
        cout<<s<<endl;
        string temp = to_string(num);
        int i=0;
        while(i < s.length() && temp[i] == s[i]) i++;
        if(i == s.length()) return num;
        int idx = -1;
        for(int j=s.length()-1; j>=i; j--){
            if(temp[j] == s[i]){
                idx = j;
                break;
            }
        }
        // while(idx < s.length()-1 && temp[idx] == temp[idx+1]) idx++;
        swap(temp[i],temp[idx]);
        return stoi(temp);
    }
};

File: 670-maximum-swap/maximum-swap.py
class Solution:
    def maximumSwap(self, num: int) -> int:

        arr = list(str(num))
        n = len(arr)
        suffix_arr = [0] * n
        suffix_arr[n-1] = n-1

        for i in range(n-2, -1, -1):
            if int(arr[i]) <= int(arr[suffix_arr[i+1]]):
                suffix_arr[i] = suffix_arr[i+1]
            else:
                suffix_arr[i] = i
        
        for i in range(len(arr)):
            if int(arr[i]) < int(arr[suffix_arr[i]]):
                arr[i], arr[suffix_arr[i]] = arr[suffix_arr[i]], arr[i]
                break
        
        return int(''.join(arr))
        

        